### ISSUE TITLE
Standardize extension points' Priority property values and policy

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#525]
Each of `IArgumentValueFormatter`, `IDummyFactory`, and `IFakeOptionsBuilder` has an integer `Priority` property that's used to choose the appropriate class when more than one such extension point applies to a particular argument/Dummy/Fake type.

The higher `Priority` wins.

`FakeConfigurator` and  `DummyFactory` return `0` as a default value, but `ArgumentValueFormatter` returns `int.MinValue`.

Currently, only `IArgumentValueFormatter` provides concrete implementations which are used by FakeItEasy, but with recent increases in the power of `IDummyFactory` and `IFakeOptionsBuilder`, we may want to implement some of FakeItEasy's features using these extension points.

To avoid confusion, and collisions with user's implementations, I propose that:

1. `FakeOptionsBuilder`, `DummyFactory`, and `ArgumentValueFormatter` all return `0` as default values for `Priority` (users can always override if they need to)
2. FakeItEasy's concrete implementations of the interfaces return very low negative values, in the `int.MinValue` range. **update: we decided on -1**
3. We document the default value of the abstract helper classes and reserve all negative `Priority` values for FakeItEasy's internal use.

**Comment by adamralph** [comment]
Sounds good to me!

For the concrete implementation values, the best possible future proofing would probably be to bisect the negative range, e.g. for 3 current implementations:

* `Impl1.Priority == int.MinValue * 1 / 4`
* `Impl2.Priority == int.MinValue * 2 / 4`
* `Impl3.Priority == int.MinValue * 3 / 4`

**Comment by blairconrad** [comment]
I'm slightly confused. Or maybe think that the scheme is not needed.
It seems like you're interested in leaving space between the priorities, yes? I imagine any concrete implementations that we provide will not have conflicts with others, so the need for a fancy strategy is quite low. I think everything we have can just use `int.MinValue`.

**Comment by adamralph** [comment]
Ah, I guess it wouldn't make sense for us to provide more than one since the higher prio would always win. In that case, yes, we can just use `int.MinValue`.

**Comment by blairconrad** [comment]
Yeah. I just wanted to reserve all negative values because I'm greedy!

**Comment by adamralph** [comment]
If we want to force everyone to use non-negative values, then should we make our priorities -1?

**Comment by blairconrad** [comment]
Oh. Crafty. We can always adjust ours downward as we need to, but if they start using negative numbers they won't see their desired behaviour.
You're an evil genius!

**Comment by adamralph** [comment]
:japanese_goblin:

**Comment by blairconrad** [comment]
It occurs to me that we could programmatically force externally-implemented extension points to use non-negative priorities by throwing whenever we encounter an offending factory/options builder/formatter. I do like the idea of backing up our "reserved for FakeItEasy" tough talk.

**Comment by adamralph** [comment]
Why is this one slated for 2.0? Is it breaking?

**Comment by adamralph** [comment]
And btw - yes, if we do make our priorities -1, I think throwing when we see some other impl. using -1 or lower is a good idea, otherwise, silent failure.

**Comment by blairconrad** [comment]
Yes, breaking if someone had had a custom priority less than 0. We'd be mucking up which extension point was found. I should've had the tag on there all along.

**Comment by thomaslevesque** [comment]
> And btw - yes, if we do make our priorities -1, I think throwing when we see some other impl. using -1 or lower is a good idea, otherwise, silent failure.

Sounds reasonable. However if we want to change that, we probably need to do it the base class (e.g. `FakeConfigurator`, so we need to make `Priority` non-virtual and introduce a virtual method or property that would be called by `Priority`.

**Comment by blairconrad** [comment]
I'd imagined performing the check as we're importing all the types (which begins in `ImportsModule.RegisterDependencies`).
I'd been thinking that we wanted to allow users to continue to create extension points just by implementing our interfaces, rather than forcing them to extend a class. Had the major .NET languages supported multiple inheritance, I wouldn't worry at all…

**Comment by thomaslevesque** [comment]
> I'd been thinking that we wanted to allow users to continue to create extension points just by implementing our interfaces, rather than forcing them to extend a class.

Good point, I forgot about that. But then how do you decide which types are allowed to have a negative priority? Just a white-list of known internal types?

**Comment by blairconrad** [comment]
Types that FakeItEasy defines. I'd check the type's assembly.

**Comment by adamralph** [comment]
> Types that FakeItEasy defines. I'd check the type's assembly.

:+1:

**Comment by thomaslevesque** [comment]
> Types that FakeItEasy defines. I'd check the type's assembly.

Sounds reasonable ;)

**Comment by blairconrad** [comment]
Rather than throwing if someone uses a negative Priority, perhaps we should stop the potential misusers from committing an error, rather than catching them after.
We could make `Priority` some type that users can't make bad values out of.
Safest, and probably easiest to understand would be to make a `Priority` a  type, like this:

```c#
public class Priority: IComparable<Priority>,  IEquatable<Priority>
{
    private readonly int value;
 
    public Priority(ushort value) { this.value = value; }
    internal Priority(int value) { this.value = value; }

    // some stuff to do with comparison and whatnot
    // maybe some implicit conversion if we want to be really fancy
}
```

Alternatively, I suppose `Priority` could be a `ushort` and we could rig our priority comparer to negate the priorities of FakeItEasy-defined types. Looks a little simpler on the outside, but the internal logic would be a little icky…

**Comment by thomaslevesque** [comment]
The `Priority` class is probably the cleanest approach

**Comment by blairconrad** [comment]
Woo hoo. It sounds like fun.

**Comment by adamralph** [comment]
The only problem with signed types is that (IIRC) they are not CLS compliant.

**Comment by blairconrad** [comment]
> The only problem with signed types is that (IIRC) they are not CLS compliant.

Yes. I hit this a few minutes before leaving the house today and would've commented, but work's been, well, work.

I'll come back with proposals later.

**Comment by thomaslevesque** [comment]
> The only problem with signed types is that (IIRC) they are not CLS compliant.

I guess you meant _unsigned_ types ;)

I'm not sure how big an issue this is. C#, VB.NET and F# all support unsigned types, but I guess there might be other, less well known .NET languages that don't.

Anyway, another option is to make the priority a `byte`, which is CLS compliant. Sure, it has a much smaller range, but I don't think there are many scenarios where you would need more than 256 priority levels... I never needed more than one myself.

**Comment by blairconrad** [comment]
Yeah, I was giving @adamralph a break on the signed/unsigned. Although I _do_ like seeing him nitpicked!

I had your thoughts about C#/VB.NET/F# as well (I had to look up the F# support to be sure), but worry a little bit about alienating some other .NET language-users. [Boo](https://en.wikipedia.org/wiki/Boo_(programming_language))?

I would be more than content using a `byte`. As you say, "256 different priorities should be enough for anyone". As you say, I expect 2 different priorities will more than satisfy most people's needs.

Joke: We could use `bool`.

**Comment by adamralph** [comment]
Sorry, yes, *un*signed types :grin: 

I don't think it's a huge problem, but I wanted to make sure we'd considered it.

I also like the `byte` proposal, and the `Priority` type proposal in general, it's an excellent idea. Perhaps it should be a `struct`?

**Comment by thomaslevesque** [comment]
> Perhaps it should be a struct?

Yes, probably.

**Comment by blairconrad** [comment]
Sorry, yes. Already was here. Thanks for the early code review!

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
